### ISSUE TITLE
Improve error handling in `check-annotations`

### ISSUE THREAD

**Issue by kairstenfay** [nextstrain/ncov-ingest#47]
@lmoncla and I currently ran into an error raised by `check-annotations` with the following output:
```
The following strains have a division_exposure but not a country_exposure.
This is ok if the country_exposure happens to match the country, but problems arise if this is not the case.
To make this clear, country_exposure should be explicitly set for the following:
strain
France/CIBU200107/2020    NaN
Name: strain, dtype: object
strains have a division_exposure but not a country_exposure
```

France/CIBU200107/2020 was correctly annotated for sampling and exposure locations. It had `division` and `division_exposure` set but not `country` or `country_exposure` annotated. However, in the metadata, the country and country exposure were France.

This issue was resolved by removing `division_exposure` annotations for France/CIBU200107/2020. If it's the case where we don't want to have a `division` and `division_exposure` annotation without a `country` or `country_exposure` annotation, then we should describe that the `division_exposure` can be dropped in the error message sent to the user.

**Comment by kairstenfay** [comment]
Another thought: would it be possible, without creating a circular dependency, to read from the metadata to determine if an annotated strain has an appropriately filled `country`, `country_exposure`, `region`, or `region_exposure`? Another issue popped up with additional sequences annotated with country exposures but not region exposures, because the region is the same in the metadata. Having to clarify in `annotations.tsv` that the region for these hasn't changed ends up creating additional work for the Nextstrain team.

**Comment by kairstenfay** [comment]
One more thought: the `check-annotations` script currently exits after the first error is encountered. Consider collecting all errors that arise from the `annotations.tsv` and providing all of them to the user at the end of runtime.

**Comment by jstoja** [comment]
I've just seen that the check caused issues on master. Sorry to read this.
I'll take a look at your findings here and make sure that the PR correctly pass CI before merging.

**Comment by jstoja** [comment]
> If it's the case where we don't want to have a division and division_exposure annotation without a country or country_exposure annotation, then we should describe that the division_exposure can be dropped in the error message sent to the user.

@jameshadfield Would you be able to confirm this? You mention auspice build errors in #39 

> would it be possible, without creating a circular dependency, to read from the metadata to determine if an annotated strain has an appropriately filled country, country_exposure, region, or region_exposure?

The annotations are used to generate the metadata with `bin/transform`. https://github.com/nextstrain/ncov-ingest/blob/master/bin/transform#L217
Therefore, I'm not sure which one should override the other one.

Although, we could use the metadata to fill the missing data from annotations, and then, the metadata would be overridden with data from annotations.
This would result essentially with the medatada generated by `bin/transform`, being a merge of the 2 files, with the annotations taking precedence in case of duplicate.

This raise the question, should we check the annotations or the metadata (just after the transform step)?

**Comment by jameshadfield** [comment]
Due to the way we perform regional subsampling, it's important that if a `division_exposure` is set in the annotations, then the `country_exposure` and `region_exposure` matches it. Note that If there are no entries specifying `<resolution>_exposure` then it'll be set to `<resolution>`.  

Imagine the following hypothetical situation:
```
region = north-america
country = USA
division = WA
```
If we set `division_exposure` to `CT` (via the annotations) then the `country_exposure` is USA and `region_exposure` is north-america, both of which match `CT` so there is no problem.

However if we set `division_exposure` to `Auckland` (via the annotations)  then this is problematic as `country_exposure` is USA and `region_exposure` is north-america which are incorrect (they should be "New Zealand" and "Oceania", respectively). 

In the latter example, we do not want to drop the `division_exposure` as it's important for our scientific understanding. Rather, we want to ensure that `country_exposure` and `region_exposure` are correct. 

When I wrote the script to track these down & fix the errors (f3d6e45974636e01950a05f032916455a956df6c), I opted for a safe but verbose solution of mandating that if `<resolution>_exposure` is set in the annotations then all coarser resolutions must also be explicitly set. There are other solutions which are less verbose but more complex.

**Comment by kairstenfay** [comment]
Thanks @jameshadfield for providing this helpful context. These tests that `check-annotations` is performing make sense to me now. 

I'm thinking about enforcing logical location hierarchies (country, division, and region) against a hierarchical structure derived from our existing `metadata.tsv`. If a `<resolution>` or `<resolution>_exposure` hierarchy is incorrect, then it would be flagged in `transform`. Do you think this is an acceptable replacement for enforcing that `division_exposure` annotations require a `country_exposure`, etc.?

**Comment by lmoncla** [comment]
Thanks @kairstenfay and @jameshadfield. The divisions error popped up because there was a sequence whose `division` was misspelled. When someone fixed the `division` spelling, they also made a correction for `division_exposure` in a separate line in the annotations file. The country was correct, so no changes were added. So in this case, I just want to confirm that the correct protocol is just to annotate a change for `division`. 

For the country error, I'm still a little confused about the proper protocol for the scenario in which we have a `country_exposure` that is in the same region. For example, we have `country = Italy`, and `country_exposure = Spain`. In this case, my impression was that we should only need to annotate `country_exposure`. Do we also then need to annotate `region_exposure`?

**Comment by jameshadfield** [comment]
> [@kairstenfay] Do you think this is an acceptable replacement for enforcing that division_exposure annotations require a country_exposure annotation too, etc.?

Not a replacement (what if the new exposure value isn't in the existing-metadata-derived hierarchy?) but I can see that this approach would be more more user friendly. It's a trade-off (as always!)

> [@lmoncla] Do we also then need to annotate region_exposure?

In your example you don't. But there were examples where the `region_exposure` needed to be specified (as it differed from the `region`) but it wasn't. This wasn't identified until the auspice builds were made, and had an element or randomness to it due to the subsampling. This is why I originally wrote the script and corrected the TSV (see the changes in f3d6e45). The script I wrote would enforce that you specified `region_exposure -> europe` as it followed a "safe rather than sorry" approach.

**Comment by kairstenfay** [comment]
> > [@kairstenfay] Do you think this is an acceptable replacement for enforcing that division_exposure annotations require a country_exposure annotation too, etc.?
> 
> Not a replacement (what if the new exposure value isn't in the existing-metadata-derived hierarchy?) but I can see that this approach would be more more user friendly. It's a trade-off (as always!)

That makes sense @jameshadfield. The goal here is definitely user friendliness. You raise a good point about how we would handle new hierarchies. I think the options for creating a hierarchy are:
1. create an exhaustive list of all possible hierarchies (not derived from metadata) from the start. This may be unfeasible.
2. maintain a list of hierarchies derived from all Nextstrain builds, reducing the frequency of times we would encounter a new hierarchy
3. create the list of hierarchies derived from only the ncov build

Following 2 or 3, once we find a new hierarchy, we could: 
a. Have some sort of hybrid approach between the logic in `check-annotations` and the location hierarchy validation. For example, a new hierarchy with annotated `division_exposure` must also have an annotation for `country_exposure` (i.e. the current enforcement would apply)
b. Assume all new hierarchies are valid until proven otherwise.

Thoughts here?

**Comment by jameshadfield** [comment]
I think that just checking against the nCoV-metadata-hierarchy and enforcing that coarser resolutions are explicitly added for novel values would be a good mix between user-friendliness and simplicity, but the decision is yours!

**Comment by kairstenfay** [comment]
Thanks @jameshadfield. I really appreciate your feedback here.

**Comment by kairstenfay** [comment]
Well, this took awhile to finally figure out how we wanted it implemented, but #54 created a new file, `./bin/check-locations`, that performs some location hierarchy validations (i.e. checks for previously unseen region-country-division-location combinations) _after_ all annotations were applied to the metadata. It replaced some of the checks in `./bin/check-annotations` that were causing the CI tests to break because a division was defined but not a country, etc., as described in the first comment of this thread. (The solution proposed by #54 was to send Slack messages on new location hierarchies that appear, prompting build maintainers to manually edit `./source-data/location_hierarchy.tsv`, rather than stopping the build.)

**Comment by kairstenfay** [comment]
I think the core of this issue has been resolved by #54, and now #39 deserves to be revisited.

Note that #39 was originally resolved by #40 (thank you @jstoja), and then I pushed [this commit](https://github.com/nextstrain/ncov-ingest/commit/8d9cbe336b65019caeb99c729ee86d4c939cf0a6) to remove `./bin/check-annotations` from the CI while we worked out how to validate annotations in a sustainable way for nCoV build maintainers.
